File: 9177/captain/Mix.cpp
#include <iostream>
#include <string>

using namespace std;

int DP[201][201];
string str1, str2, str3;
int l1, l2, l3;

int dfs(int i, int j) {
    if (DP[i][j] != -1) return DP[i][j];
    if (i + j == l3 || (i > l1 && j >= l2)) return true;
    auto &ret = DP[i][j] = false;
    if (str1[i] == str3[i + j] && i < l1) ret |= dfs(i + 1, j);
    if (str2[j] == str3[i + j] && j < l2) ret |= dfs(i, j + 1);
    return ret;
}

int main() {
    int t;
    cin >> t;
    for (int i = 0; i < t; i++) {
        cin >> str1 >> str2 >> str3;
        for (int j = 0; j < 201; j++)
            for (int k = 0; k < 201; k++)
                DP[j][k] = -1;
        l1 = str1.size();
        l2 = str2.size();
        l3 = str3.size();
        string result = dfs(0, 0) ? "yes" : "no";
        cout << "Data set " << i + 1 << ": " << result << endl;
    }
}

File: 9177/captain/Mix.py
import sys
word1 = None
word2 = None
word3 = None
word1Length = 0
word2Length = 0
word3Length = 0
result = 'no'

def verify(index1, index2, index3):
    if index3 == word3Length:
        global result
        result = 'yes'
    if index1 != word1Length and word1[index1] == word3[index3]:
        verify(index1 + 1, index2, index3 + 1)
    if index2 != word2Length and word2[index2] == word3[index3]:
        verify(index1, index2 + 1, index3 + 1)

if __name__=='__main__':
    case = int(sys.stdin.readline())
    word = [[] for _ in range(1000)]
    for i in range(case):
        word[i] = list(map(str, sys.stdin.readline().strip().split(' ')))
    for i in range(case):
        result = 'no'
        word1, word2, word3 = map(str, word[i])
        word1Length = len(word1)
        word2Length = len(word2)
        word3Length = len(word3)
        result = verify(0, 0, 0)
        print('Data set {0}: {1}'.format(i + 1, 'yes' if result else 'no'))